designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,534 @@
+
+const config1 = {
+    name: "config1",
+    files: ["**/*.cjs.js"],
+    languageOptions: {
+        sourceType: "commonjs",
+        globals: {
+            ...globals.node
+        }
+    }
+};
+
+const config2 = {
+    name: "config2",
+    files: ["**/*.js"],
+    ignores: ["**/tests"],
+    rules: {
+        "no-console": "error"
+    }
+};
+
+
+export default [
+
+    {
+        name: "myconfig",
+        files: ["**/src/*.js"],
+        ignores: ["**/*.test.js"]
+        extends: [config1, config2],
+        rules: {
+            semi: "error"
+        }
+    }
+
+];
+```
+
+Here, the `files` keys will be combined and the `ignores` key will be merged, resulting in a final config that looks like this:
+
+```js
+
+export default [
+
+    // intersected files and original ignores
+    {
+        name: "myconfig > config1",
+        files: [["**/src/*.js", "**/*.cjs.js"]],
+        ignores: ["**/*.test.js"],
+        languageOptions: {
+            sourceType: "commonjs",
+            globals: {
+                ...globals.node
+            }
+        }
+    },
+
+    // intersected files and merged ignores
+    {
+        name: "myconfig > config2",
+        files: [["**/src/*.js", "**/*.js"]],
+        ignores: ["**/*.test.js", "**/tests"],
+        rules: {
+            "no-console": "error"
+        }
+    },
+
+    // original config
+    {
+        name: "myconfig",
+        files: ["**/src/*.js"],
+        ignores: ["**/*.test.js"]
+        rules: {
+            semi: "error"
+        }
+    }
+
+];
+```
+
+When an extended config and the base config both have multiple `files` entries, then the result is a `files` entry containing all combinations. For example:
+
+```js
+export default [
+    {
+        files: ["src/**", "lib/**"],
+        extends: [{ files: ["**/*.js", "**/*.mjs"] }]
+    }
+];
+```
+
+This config would be calculated as the following:
+
+```js
+export default [
+    {
+        files: [
+            ["src/**", "**/*.js"],
+            ["lib/**", "**/*.js"],
+            ["src/**", "**/*.mjs"],
+            ["lib/**", "**/*.mjs"]
+        ]
+    }
+];
+```
+
+Notes:
+
+1. The `files` and `ignores` values from the base config always come first in the calculated `files` and `ignores`. If there's not a match, then there's no point in continuing on to use the patterns from the extended configs.
+1. The `name` key is generated for calculated config objects so that it indicates the inheritance from another config. The extended configs don't exist in the final representation of the config array.
+1. These behaviors differ from the [typescript-eslint extends helper](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/typescript-eslint/src/config-helper.ts). The `tsconfig` helper overwrites the `files` and `ignores` fields on any config contained in `extends` whereas this proposal creates combination `files` and `ignores`, merging the base config with the extended configs.
```

Comment:
```suggestion
1. These behaviors differ from the [typescript-eslint extends helper](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/typescript-eslint/src/config-helper.ts). The typescript-eslint `config` helper overwrites the `files` and `ignores` fields on any config contained in `extends` whereas this proposal creates combination `files` and `ignores`, merging the base config with the extended configs.
```

tsconfig is something else 🙂

Reply:
Heh thanks. Last day before break and my brain was mush.